File: src/lib/calendar/EventScheduler.ts
/**
 * "Algoritmos" de alto nível que combinam EventList + CalendarGrid.
 *
 * Conceitos AP CS A:
 *  - Unit 4 + 7 (Iteration + ArrayList): traverse, busca linear, contagem.
 *  - Unit 7 (Sorting): demonstra Insertion Sort por data, alternativa
 *    à Selection Sort que está em EventList.
 *  - Unit 10 (Recursion): countEventsRecursive faz a contagem de eventos
 *    recursivamente, espelhando o estilo dos exercícios de FRQ recursivos.
 *  - Unit 5 (Static methods): tudo aqui é estático — esta é uma
 *    "utility class" no estilo Math do Java.
 */

import { CalendarEvent } from "./CalendarEvent"
import { EventList } from "./EventList"

export class EventScheduler {
  // Unit 7: Insertion Sort — outro algoritmo do CED.
  // Cada iteração pega o elemento da posição i e o "encaixa" na sub-lista
  // ordenada à esquerda.
  public static insertionSortByDate(list: EventList): void {
    const n = list.size()
    for (let i = 1; i < n; i++) {
      const key = list.get(i)
      let j = i - 1
      while (j >= 0 && list.get(j).compareDate(key) > 0) {
        list.set(j + 1, list.get(j))
        j--
      }
      list.set(j + 1, key)
    }
  }

  // Unit 10: Recursão — conta quantos eventos existem na lista a partir
  // do índice `start`. Caso base: índice fora do array. Caso recursivo:
  // 1 + chamada para o índice seguinte. É um clássico de FRQ.
  public static countEventsRecursive(
    list: EventList,
    start = 0,
  ): number {
    if (start >= list.size()) return 0
    return 1 + EventScheduler.countEventsRecursive(list, start + 1)
  }

  // Unit 10: Recursão com filtro — conta quantos eventos caem em uma
  // categoria específica. Caso base + chamada com índice seguinte
  // somando 0 ou 1 dependendo da condição.
  public static countByCategoryRecursive(
    list: EventList,
    category: CalendarEvent["getCategory"] extends () => infer R ? R : never,
    start = 0,
  ): number {
    if (start >= list.size()) return 0
    const head = list.get(start).getCategory() === category ? 1 : 0
    return head + EventScheduler.countByCategoryRecursive(list, category, start + 1)
  }

  // Unit 4: traverse linear para encontrar o próximo evento futuro.
  // Pré-condição: lista ordenada por data.
  public static nextUpcoming(
    sorted: EventList,
    now: Date = new Date(),
  ): CalendarEvent | null {
    const t = now.getTime()
    for (let i = 0; i < sorted.size(); i++) {
      const ev = sorted.get(i)
      if (ev.getDate().getTime() >= t) return ev
    }
    return null
  }

  // Unit 4 + 7: agrupa eventos por dia para a renderização do mês.
  // Retorna um Map<chave do dia, EventList>.
  public static groupByDay(list: EventList): Map<string, EventList> {
    const map = new Map<string, EventList>()
    for (const ev of list.toArray()) {
      const d = ev.getDate()
      const key = `${d.getFullYear()}-${d.getMonth()}-${d.getDate()}`
      if (!map.has(key)) map.set(key, new EventList())
      map.get(key)!.add(ev)
    }
    return map
  }

  // Helper: chave consistente para indexar Map<string, EventList>.
  public static dayKey(year: number, month: number, day: number): string {
    return `${year}-${month}-${day}`
  }

  // ───────── HEATMAP / agregações 2D (Unit 8) ─────────

  // Unit 8: array 2D de contagens — mesmo formato da CalendarGrid (6 semanas
  // × 7 dias). Cada célula recebe o número de eventos daquele dia (do mês
  // corrente; spillover marcado como -1 pra UI ignorar).
  public static heatmapMatrix(
    list: EventList,
    grid: {
      getCells(): {
        year: number
        month: number
        day: number
        inMonth: boolean
      }[][]
    },
  ): number[][] {
    const buckets = EventScheduler.groupByDay(list)
    const cells = grid.getCells()
    const matrix: number[][] = []
    for (let r = 0; r < cells.length; r++) {
      const row: number[] = []
      for (let c = 0; c < cells[r].length; c++) {
        const cell = cells[r][c]
        if (!cell.inMonth) {
          row.push(-1)
          continue
        }
        const key = EventScheduler.dayKey(cell.year, cell.month, cell.day)
        row.push(buckets.get(key)?.size() ?? 0)
      }
      matrix.push(row)
    }
    return matrix
  }

  // Unit 4 + 8: max numa matriz 2D ignorando -1.
  public static maxInMatrix(matrix: number[][]): number {
    let max = 0
    for (let r = 0; r < matrix.length; r++) {
      for (let c = 0; c < matrix[r].length; c++) {
        if (matrix[r][c] > max) max = matrix[r][c]
      }
    }
    return max
  }
}

File: java/EventScheduler.java
/*
 * EventScheduler — utility class com algoritmos sobre EventList.
 *
 * Unit 5 — todos os metodos sao static (utility class no estilo Math).
 * Unit 10 — RECURSAO: countEventsRecursive, countByCategoryRecursive.
 * Unit 8 — agregacao 2D (heatmap).
 */
import java.util.ArrayList;

public class EventScheduler {

    // ---------- Unit 10: RECURSAO ----------
    //
    // Conta quantos eventos existem na lista a partir de `start`.
    // Caso base: indice fora dos limites -> 0.
    // Caso recursivo: 1 + chamada para o proximo indice.
    public static int countEventsRecursive(EventList list, int start) {
        if (start >= list.size()) {
            return 0;  // caso base
        }
        // caso recursivo
        return 1 + countEventsRecursive(list, start + 1);
    }

    // Recursao com filtro: conta eventos de uma categoria especifica.
    public static int countByCategoryRecursive(EventList list,
                                                String category,
                                                int start) {
        if (start >= list.size()) {
            return 0;
        }
        int head = list.get(start).getCategory().equals(category) ? 1 : 0;
        return head + countByCategoryRecursive(list, category, start + 1);
    }

    // ---------- Unit 4: traverse linear ----------
    //
    // Pre-condicao: lista ordenada por data (selectionSortByDate).
    // Devolve o primeiro evento >= agora, ou null.
    public static CalendarEvent nextUpcoming(EventList sortedList,
                                              int yearNow, int monthNow,
                                              int dayNow, int hourNow,
                                              int minuteNow) {
        int target = (yearNow * 12 + monthNow) * 31 * 24 * 60
                   + dayNow * 24 * 60 + hourNow * 60 + minuteNow;
        for (int i = 0; i < sortedList.size(); i++) {
            CalendarEvent ev = sortedList.get(i);
            int t = ev.toMinutes();
            if (t >= target) return ev;
        }
        return null;
    }

    // ---------- Unit 8 — heatmap (agregacao 2D) ----------
    //
    // Recebe uma EventList e uma CalendarGrid. Devolve uma matriz
    // 6x7 com a contagem de eventos por celula. Celulas fora do mes
    // ganham -1 pra que a UI possa ignorar.
    public static int[][] heatmapMatrix(EventList list, CalendarGrid grid) {
        int rows = grid.getRowCount();
        int cols = grid.getColCount();
        int[][] matrix = new int[rows][cols];
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                DayCell cell = grid.getCell(r, c);
                if (!cell.isInMonth()) {
                    matrix[r][c] = -1;
                    continue;
                }
                EventList dayList = list.filterByDate(
                    cell.getYear(), cell.getMonth(), cell.getDay()
                );
                matrix[r][c] = dayList.size();
            }
        }
        return matrix;
    }

    // Unit 4 + 8: varre a matriz inteira encontrando o maior valor
    // (ignora -1 que e spillover do mes vizinho).
    public static int maxInMatrix(int[][] matrix) {
        int max = 0;
        for (int r = 0; r < matrix.length; r++) {
            for (int c = 0; c < matrix[r].length; c++) {
                if (matrix[r][c] > max) {
                    max = matrix[r][c];
                }
            }
        }
        return max;
    }

    // ---------- Helper: junta varios eventos em um ArrayList -----
    //
    // Agrupa eventos por dia. Equivalente a Map<String,EventList> em TS.
    // Aqui usamos ArrayList simples por chave construida na hora.
    public static ArrayList<CalendarEvent> eventsForDay(
            EventList list, int year, int month, int day) {
        ArrayList<CalendarEvent> out = new ArrayList<CalendarEvent>();
        for (int i = 0; i < list.size(); i++) {
            CalendarEvent ev = list.get(i);
            if (ev.isOnDate(year, month, day)) {
                out.add(ev);
            }
        }
        return out;
    }
}
